### ISSUE TITLE
Journal Code Review - unable to reproduce human notebook

### ISSUE THREAD

**Issue by mathieuboudreau** [spinal-cord-7t/coil-qc-code#161]
We received the feedback that the code reviewer for the journal had issues reproducing/running the human notebook. This issue will be to document my own attempts and how we can improve documentation / installation instructions (i.e. maybe more detailed python / os versions).

**Comment by mathieuboudreau** [comment]
I followed the installation instructions on my Mac OS (python 3.11.0, conda vent) and ran the human notebook. 

It took ~ 6 hours to run on my Mac. At first glance, all the figures looked to be reproduced relative to our current jupyter book build (https://spinal-cord-7t.github.io/coil-qc-code/README.html), except one of the figures was missing a line on the plot:

# My run

![Image](https://github.com/user-attachments/assets/7bbdc0ed-55de-4627-9195-d7eaddcfcf2c)

# Jupyter Book

![Image](https://github.com/user-attachments/assets/ab62b74f-59ee-4a1b-98a4-1d6817809429)

For some reason, the average wasn't plotted for MSSM (far right). Not a big issue, just thought I'd note it here. 

I'm going to re-run it now on github actions, and do so for all three OS's to ensure there wasn't an incompatibility issue that the reviewer encountered

**Comment by evaalonsoortiz** [comment]
Both subject 3's line and the average line look to be missing

**Comment by mathieuboudreau** [comment]
> Both subject 3's line and the average line look to be missing

Good catch!

I just reran the analysis on Actions/Linux (), and it replicates this same issue I first saw in my local Mac build. I'll have to look at the logs to see if anything pops up

**Comment by mathieuboudreau** [comment]
Debugging a bit, subject 3 of site MSSM's SNR_90 data is all NaN

<img width="930" alt="Image" src="https://github.com/user-attachments/assets/949c8364-9240-40d6-80aa-9562a9927ed2" />

tbt why

**Comment by mathieuboudreau** [comment]
Here's a printed log of when this subject's data is loaded:

```
20
18
/Users/mathieuboudreau/neuropoly/github/coil-qc-code/data-human/sub-MSSM3/fmap
Initialize list to collect data for this subject
/Users/mathieuboudreau/neuropoly/github/coil-qc-code/data-human/derivatives/results/sub-MSSM3_SNRmap.csv
0     45.676768
1     60.520492
2     76.732000
3     85.598214
4     94.643172
5    100.567568
6    105.291291
7    107.933921
8           NaN
Name: WA(), dtype: float64
Compute stats on the non-resampled data (to avoid interpolation errors)
84.62042814195408
20.87513218966155
Normalize the x-axis to a 1-0 scale for each subject (to go from superior-inferior direction)
[1.    0.875 0.75  0.625 0.5   0.375 0.25  0.125 0.   ]
Interpolate to the fixed grid
[         nan          nan          nan          nan          nan
          nan          nan          nan          nan          nan
          nan          nan          nan 107.80045457 107.58690876
 107.37336295 107.15981714 106.94627133 106.73272552 106.51917971
 106.30563389 106.09208808 105.87854227 105.66499646 105.45145065
 105.19586253 104.81414748 104.43243243 104.05071738 103.66900234
 103.28728729 102.90557224 102.52385719 102.14214214 101.76042709
 101.37871205 100.996997   100.61528195 100.1486709   99.66993185
  99.19119279  98.71245374  98.23371469  97.75497564  97.27623659
  96.79749754  96.31875849  95.84001943  95.36128038  94.88254133
  94.27771898  93.54681332  92.81590766  92.085002    91.35409634
  90.62319069  89.89228503  89.16137937  88.43047371  87.69956805
  86.96866239  86.23775674  85.50865657  84.79219481  84.07573304
  83.35927128  82.64280952  81.92634776  81.209886    80.49342424
  79.77696248  79.06050072  78.34403896  77.6275772   76.91111544
  75.74948435  74.43946349  73.12944262  71.81942176  70.50940089
  69.19938003  67.88935917  66.5793383   65.26931744  63.95929657
  62.64927571  61.33925484  60.07068198  58.87118912  57.67169626
  56.47220341  55.27271055  54.07321769  52.87372483  51.67423197
  50.47473911  49.27524625  48.07575339  46.87626054  45.67676768]
Apply smoothing
[nan nan nan nan nan nan nan nan nan nan nan nan nan nan nan nan nan nan
 nan nan nan nan nan nan nan nan nan nan nan nan nan nan nan nan nan nan
 nan nan nan nan nan nan nan nan nan nan nan nan nan nan nan nan nan nan
 nan nan nan nan nan nan nan nan nan nan nan nan nan nan nan nan nan nan
 nan nan nan nan nan nan nan nan nan nan nan nan nan nan nan nan nan nan
 nan nan nan nan nan nan nan nan nan nan]
[array([nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan,
       nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan,
       nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan,
       nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan,
       nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan,
       nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan,
       nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan,
       nan, nan, nan, nan, nan, nan, nan, nan, nan])]
```

**Comment by mathieuboudreau** [comment]
sub-MSSM3_SNRmap.csv seems to be missing WA data in the first vert level, and this translated into a NaN after loading the file which just propagates,

<img width="885" alt="Image" src="https://github.com/user-attachments/assets/0c791a33-b7b9-4f34-924d-dc5f3e0410e3" />

**Comment by mathieuboudreau** [comment]
Looks like the cause is that the segmentation for `sub-MSSM3_acq-coilQaSagLarge_SNR_T0000` doesn't reach up to what was labelled as the first vert level,

Greyscale - image
Color label - labels
White - segmentation

I'm not quite sure 1) what caused this (SCT version? dependency?) and 2) if it could have impacted other results (visually, it doesn't appear so). @jcohenadad who from SCT could I reach out to debug this a bit more? @joshuacwnewton or someone else?

**Comment by jcohenadad** [comment]
@mathieuboudreau I'll look into it

**Comment by joshuacwnewton** [comment]
> I'm not quite sure 1) what caused this (SCT version? dependency?) and 2) if it could have impacted other results (visually, it doesn't appear so). [@jcohenadad](https://github.com/jcohenadad) who from SCT could I reach out to debug this a bit more? [@joshuacwnewton](https://github.com/joshuacwnewton) or someone else?

Just for clarity, the notebook uses the contrast agnostic task, right?

https://github.com/spinal-cord-7t/coil-qc-code/blob/6f70b7c4ceec9e3351d75d1ab94529aab836dcb3/data_processing-human.ipynb#L331

The underlying model has changed a few times depending on the SCT version (see [these releases](https://github.com/sct-pipeline/contrast-agnostic-softseg-spinalcord/releases) and the [corresponding SCT commits](https://github.com/spinalcordtoolbox/spinalcordtoolbox/commits/master/spinalcordtoolbox/deepseg/models.py)).

I notice that this repo doesn't specify the exact version of SCT used (it just says "Install [Spinal Cord Toolbox](https://spinalcordtoolbox.com/user_section/installation.html)"). The preprint also doesn't specify the exact version of SCT used:

![Image](https://github.com/user-attachments/assets/76f82c77-1f90-42ea-90aa-0b5e506cb4bf)

So, is it possible that the notebook was developed using an earlier version of SCT, but the reviewer installed the latest version?

**Comment by mathieuboudreau** [comment]
>So, is it possible that the notebook was developed using an earlier version of SCT, but the reviewer installed the latest version? 

Yep! So, next step for me will be to roll back to whatever the latest version was ~ Nov 4 2024. But also, maybe the notebook should be edited such that if there is missing WA data in one of the verts, it's not counted in the mean as  that is what resulted in the NaN's and missing plots.

**Comment by mathieuboudreau** [comment]
Ok, pinned ST to v6.4 and all the requirements to `<=` the version available on ~ Nov 5th, running here and will report back if this resolved it.

https://github.com/spinal-cord-7t/coil-qc-code/actions/runs/13927683366

In a more abstract way, I'd be curious to know if you believe that the current v6.5 segmentation models are better than the v6.4 / why it's different enough that there was an issue here that we encountered.

**Comment by joshuacwnewton** [comment]
> In a more abstract way, I'd be curious to know if you believe that the current v6.5 segmentation models are better than the v6.4 / why it's different enough that there was an issue here that we encountered.

Typically the differences between contrast-agnostic models include adding additional contrasts/datasets to the pool of training data, to widen the types of images the model can be applied to. But, in this specific case, there was a larger change:

- v6.4 (contrast-agnostic [v2.4](https://github.com/sct-pipeline/contrast-agnostic-softseg-spinalcord/releases/tag/v2.4)): Trained using MONAI framework
- v6.5 (contrast-agnostic [v3.1](https://github.com/sct-pipeline/contrast-agnostic-softseg-spinalcord/releases/tag/v3.1)): Trained using nnUNetv2 framework. 
  From the releases notes:
  > Replacing monai-based model with a 3D nnUNet model trained from scratch. Tested on a wide variety of pathologies and contrasts. Works especially well on compressed spinal cords, fixes issues with shifted predictions (consistent in monai-based models)

Tagging @naga-karthik just in case he has more to add here.

**Comment by jcohenadad** [comment]
Thank you @joshuacwnewton. I'm currently trying to reproduce the error locally, but the bottomline is that we absolutely should fix SCT version for this project.

**Comment by naga-karthik** [comment]
Thanks for the tag @joshuacwnewton ! I think you summarized it well. My suspicion is also about the discrepancies in the model versions that could have caused the issue. We had a several updates in a short timeframe

> In a more abstract way, I'd be curious to know if you believe that the current v6.5 segmentation models are better than the v6.4

@mathieuboudreau Generally, with each new version the model segmentations improved at least on the test sets I have evaluated on. But between SCT v6.4 and v6.5, I wouldn't consider switching to the new model for your notebook -- i.e. the performance of the model in SCT v6.4 should be similar to the one in SCT v6.5, so you can continue with the version that you used in ~ Nov 2024. 

However, if you plan to do some follow-up analysis (outside the scope of the current notebook), please consider switching (and fixing) to SCT v7.0 as this model is more capable than the previous versions.

**Comment by mathieuboudreau** [comment]
Hum, now in this branch (https://github.com/spinal-cord-7t/coil-qc-code/tree/mb/pin), this run(https://github.com/spinal-cord-7t/coil-qc-code/actions/runs/13927683366) with SCT pinned to 6.4 & the reqs pinned to <= versions @ early november actually resulted in more missing lined (assuming these are due to NaNs but have other checked).

<img width="710" alt="Image" src="https://github.com/user-attachments/assets/ec1ed302-5b8c-4998-9c48-f040da7c9c43" />

I wasn't expecting this; I think the next thing I should try maybe is pinning the commit and not version that was on there around early nov; however, if for examples the models are modified in a different repo and the latest one is downloaded, maybe it's not fetching the version that was originally used? Also, why is this only impacting the SNR_90 images and not the other metrics above....

@jcohenadad thoughts on what direction I should be focusing on? Exploring if this is a version issue, or patching the notebook so that the mean doesn't consider the missing WA value for that one vertex (I've got a feeling that the figure plot will be similar to the previous version, just not identical maybe)

-------

Off topic a bit, but @jcohenadad this is one example where where a NeuroLibre preprint could have helped, as there would have been a saved docker container archived with all the versions from the initial run + the data would have been archived on their servers too.

**Comment by mathieuboudreau** [comment]
Always subject 3, too.

**Comment by mathieuboudreau** [comment]
> > [@jcohenadad](https://github.com/jcohenadad) thoughts on what direction I should be focusing on? Exploring if this is a version issue, or patching the notebook so that the mean doesn't consider the missing WA value for that one vertex (I've got a feeling that the figure plot will be similar to the previous version, just not identical maybe)
> 
> i would explore the version issue-- essentially checking when we published the book (at that time, it worked) and see what SCT version that was relating to

Oh sorry, thats what I did already for my comment above. I’ll look into maybe using the exact commit and not version, as we were cloning from master

**Comment by mathieuboudreau** [comment]
Pinning to SCT v6.5 worked!

![Image](https://github.com/user-attachments/assets/8a314d01-95a2-4506-a5c7-674356e6620d)

The necessary changes there must have been in master while we were git cloning in nov-jan, just not released yet. 

I'm going to move ahead with documenting these necessary steps to reproduce the original book (including #162), and update the readme accordingly

**Comment by evaalonsoortiz** [comment]
Thank you Mathieu!
